walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,84 @@
 
 package com.google.sps;
 
+import com.google.common.collect.Sets;
+import java.util.Set;
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.stream.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+
+    if (getAllAttendees(request).isEmpty() || events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    Collection<TimeRange> freeTimes = getFreeTimes(events, request.getDuration(), getAllAttendees(request));
+
+    if (freeTimes.isEmpty()) {
+        freeTimes = getFreeTimes(events, request.getDuration(), request.getAttendees());
+    }
+
+    return freeTimes;
+  }
+
+  public Collection<String> getAllAttendees(MeetingRequest request) {
+    Collection<String> allAttendees = new HashSet<>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    return allAttendees;
+  }
+
+  public Collection<TimeRange> getFreeTimes(Collection<Event> events, long duration, Collection<String> attendees) {
+    Collection<TimeRange> eventsWithAttendees = filterByAttendees(events, attendees);
+    Collection<TimeRange> freeTimes = splitDayByEvents(eventsWithAttendees);
+    return filterByDuration(freeTimes, duration);
+  }
+
+  public Collection<TimeRange> filterByAttendees(Collection<Event> events, Collection<String> attendees) {
+    return events.stream().filter(e -> (!Collections.disjoint(e.getAttendees(), attendees))).map(e -> e.getWhen()).collect(Collectors.toList());
+  }
+
+  public Collection<TimeRange> filterByDuration(Collection<TimeRange> times, long duration) {
+    return times.stream().filter(c -> c.duration() >= duration).collect(Collectors.toList());
+  }
+
+  public Collection<TimeRange> splitDayByEvents(Collection<TimeRange> times) {
```

Comment:
The parameter is called `times` but is a list of time ranges, not individual times. this might be confusing. Optional: re-name it to time_ranges.